src/k8s_api_client_watch.erl

```diff
@@ -106,6 +119,20 @@ handle_event(info, {gun_response, ConnPid, StreamRef, fin, Status, _Headers}, _,
     ?LOG(debug, "~p: stream closed with status ~p", [ConnPid, Status]),
     {stop, normal};
 
+handle_event(info, {gun_data, ConnPid, StreamRef, nofin, _Bin} = Msg, {_Phase, error} = State,
+	     #{conn := ConnPid, stream := StreamRef} = _Data0) ->
+    ?LOG(critical, "ErrorEvMsg: ~p, State: ~p", [Msg, State]),
+    keep_state_and_data;
+
+handle_event(info, {gun_data, ConnPid, StreamRef, fin, _Bin} = Msg, {_Phase, error} = State,
+	     #{conn := ConnPid, stream := StreamRef} = _Data0) ->
+    ?LOG(critical, "ErrorEvMsg: ~p, State: ~p", [Msg, State]),
+
+    %% the normal supervisor restarts will deal with connection restarts, but here we
+    %% are in a critical error state, kill the application to prevent supervisor restarts
+    application:stop(k8s_api_client),
```

Comment:
I don't like the ide of a lower-level component killing the top level component... Maybe a 0 threshold for restarts would be enough to shut down the application?

But as alpha, I guess we can leave it like that and check it out later (in that case, just add a TODO line).

Reply:
me neither, but it's to only sensible quick solution that I've found.

The watch process is monitored a "simple_one_for_one" supervisor and terminating the clients themself is not always a critical failure. A simple connection close and reconnect will also terminate and restart the client (just like the comment says).

The error state can only be reached from a condition that we can not deal with and where a simple connection restart will solve the problem. Something "harder" is therefore needed. The supervisor does not give us any way to communicate that.